File: lib/veidrodelis/map_proj/lists.ex
defmodule Vdr.MapProj.Lists do
  @moduledoc """
  List store operations for MapProj.

  Stores list values in a map structure using raw binaries.
  Each list is stored as a plain Elixir list of binary elements.
  """

  @type db :: non_neg_integer()
  @type key :: binary()
  @type element :: binary()
  @type store :: %{db() => %{key() => list()}}
  @type position :: :before | :after

  @doc """
  Insert all values at the head of the list (left side).
  Returns the updated store.
  """
  @spec lpush(store(), db(), key(), [element()]) :: store()
  def lpush(store, db, key, values) when is_list(values) do
    db_map = Map.get(store, db, %{})
    list = get_entries(db_map, key)

    new_entries = Enum.reverse(values) ++ list

    store_entries(store, db, key, new_entries)
  end

  @doc """
  Insert all values at the tail of the list (right side).
  Returns the updated store.
  """
  @spec rpush(store(), db(), key(), [element()]) :: store()
  def rpush(store, db, key, values) when is_list(values) do
    db_map = Map.get(store, db, %{})
    list = get_entries(db_map, key)

    new_entries = list ++ values

    store_entries(store, db, key, new_entries)
  end

  @doc """
  Insert values at the head, only if the key exists.
  Returns the updated store.
  """
  @spec lpushx(store(), db(), key(), [element()]) :: store()
  def lpushx(store, db, key, values) when is_list(values) do
    db_map = Map.get(store, db, %{})

    if Map.has_key?(db_map, key) do
      lpush(store, db, key, values)
    else
      store
    end
  end

  @doc """
  Insert values at the tail, only if the key exists.
  Returns the updated store.
  """
  @spec rpushx(store(), db(), key(), [element()]) :: store()
  def rpushx(store, db, key, values) when is_list(values) do
    db_map = Map.get(store, db, %{})

    if Map.has_key?(db_map, key) do
      rpush(store, db, key, values)
    else
      store
    end
  end

  @doc """
  Remove and return the first element of the list (from the head).
  Returns the updated store.
  """
  @spec lpop(store(), db(), key()) :: store()
  def lpop(store, db, key) do
    db_map = Map.get(store, db, %{})

    case get_entries(db_map, key) do
      [] ->
        store

      [_head | tail] ->
        store_entries(store, db, key, tail)
    end
  end

  @doc """
  Remove and return the last element of the list (from the tail).
  Returns the updated store.
  """
  @spec rpop(store(), db(), key()) :: store()
  def rpop(store, db, key) do
    db_map = Map.get(store, db, %{})
    list = get_entries(db_map, key)

    case Enum.split(list, -1) do
      {[], []} ->
        store

      {init, [_last]} ->
        store_entries(store, db, key, init)
    end
  end

  @doc """
  Remove the first count occurrences of element from the list.
  Returns the updated store.
  """
  @spec lrem(store(), db(), key(), integer(), element()) :: store()
  def lrem(store, db, key, count, element) do
    db_map = Map.get(store, db, %{})
    list = get_entries(db_map, key)

    new_list =
      cond do
        count > 0 ->
          remove_n_from_head(list, element, count)

        count < 0 ->
          list
          |> Enum.reverse()
          |> remove_n_from_head(element, abs(count))
          |> Enum.reverse()

        count == 0 ->
          Enum.reject(list, fn elem -> elem == element end)
      end

    store_entries(store, db, key, new_list)
  end

  @doc """
  Trim the list to the specified range.
  Returns the updated store.
  """
  @spec ltrim(store(), db(), key(), integer(), integer()) :: store()
  def ltrim(store, db, key, start_idx, stop_idx) do
    db_map = Map.get(store, db, %{})
    list = get_entries(db_map, key)
    len = length(list)

    start_pos = normalize_index(start_idx, len)
    stop_pos = normalize_index(stop_idx, len)

    new_list =
      if start_pos > stop_pos or start_pos >= len do
        []
      else
        Enum.slice(list, start_pos..min(stop_pos, len - 1))
      end

    store_entries(store, db, key, new_list)
  end

  @doc """
  Set the list element at index to value.
  Returns the updated store.
  """
  @spec lset(store(), db(), key(), integer(), element()) :: store()
  def lset(store, db, key, index, value) do
    db_map = Map.get(store, db, %{})
    list = get_entries(db_map, key)
    len = length(list)

    pos = if index < 0, do: len + index, else: index

    if pos >= 0 and pos < len do
      new_list = Elixir.List.replace_at(list, pos, value)
      store_entries(store, db, key, new_list)
    else
      store
    end
  end

  @doc """
  Insert value before or after the pivot element.
  Returns the updated store.
  """
  @spec linsert(store(), db(), key(), position(), element(), element()) :: store()
  def linsert(store, db, key, position, pivot, value)
      when position in [:before, :after] do
    db_map = Map.get(store, db, %{})
    list = get_entries(db_map, key)

    case Enum.find_index(list, fn elem -> elem == pivot end) do
      nil ->
        store

      pivot_idx ->
        insert_idx = if position == :before, do: pivot_idx, else: pivot_idx + 1
        new_list = Elixir.List.insert_at(list, insert_idx, value)
        store_entries(store, db, key, new_list)
    end
  end

  @doc """
  Atomically pop the last element from source and push it to the head of dest.
  Returns the updated store.
  """
  @spec rpoplpush(store(), db(), key(), key()) :: store()
  def rpoplpush(store, db, source, dest) do
    db_map = Map.get(store, db, %{})
    source_list = get_entries(db_map, source)

    case Enum.split(source_list, -1) do
      {[], []} ->
        store

      {source_init, [popped]} ->
        if source == dest do
          # Same list - rotate: move last element to front
          store_entries(store, db, source, [popped | source_init])
        else
          # Different lists
          dest_list = get_entries(db_map, dest)
          store = store_entries(store, db, source, source_init)
          store_entries(store, db, dest, [popped | dest_list])
        end
    end
  end

  @doc """
  Get a range of elements from the list.
  Both start and stop are inclusive and support negative indices.
  Returns binary values.
  """
  @spec lrange(store(), db(), key(), integer(), integer()) :: [binary()]
  def lrange(store, db, key, start_idx, stop_idx) do
    db_map = Map.get(store, db, %{})
    list = get_entries(db_map, key)
    len = length(list)

    start_pos = normalize_index(start_idx, len)
    stop_pos = normalize_index(stop_idx, len)

    if start_pos > stop_pos or start_pos >= len do
      []
    else
      Enum.slice(list, start_pos..min(stop_pos, len - 1))
    end
  end

  @doc """
  Returns the length of the list.
  """
  @spec llen(store(), db(), key()) :: non_neg_integer()
  def llen(store, db, key) do
    db_map = Map.get(store, db, %{})
    list = get_entries(db_map, key)
    length(list)
  end

  # Private helpers

  @spec get_entries(%{key() => list()}, key()) :: [binary()]
  defp get_entries(db_map, key) do
    case Map.get(db_map, key) do
      nil -> []
      list when is_list(list) -> list
      _ -> []
    end
  end

  @spec store_entries(store(), db(), key(), [binary()]) :: store()
  defp store_entries(store, db, key, []) do
    # Remove the key when the list is empty
    case Map.get(store, db) do
      nil ->
        store

      db_map ->
        new_db_map = Map.delete(db_map, key)
        Map.put(store, db, new_db_map)
    end
  end

  defp store_entries(store, db, key, entries) when is_list(entries) do
    db_map = Map.get(store, db, %{})
    new_db_map = Map.put(db_map, key, entries)
    Map.put(store, db, new_db_map)
  end

  @spec remove_n_from_head([binary()], binary(), non_neg_integer()) :: [binary()]
  defp remove_n_from_head(list, _element, 0), do: list
  defp remove_n_from_head([], _element, _count), do: []

  defp remove_n_from_head([head | tail], element, count) do
    if head == element do
      remove_n_from_head(tail, element, count - 1)
    else
      [head | remove_n_from_head(tail, element, count)]
    end
  end

  @spec normalize_index(integer(), non_neg_integer()) :: non_neg_integer()
  defp normalize_index(idx, len) when idx < 0 do
    max(0, len + idx)
  end

  defp normalize_index(idx, _len) do
    max(0, idx)
  end
end

File: native/vdr_ts_nif/src/storage/lists.rs
use std::collections::BTreeMap;
use im::Vector;
use super::bytes::Bytes;
use super::types::StorageValue;
use crate::storage::StorageInner;

impl StorageInner {
    /// Push elements to the left (head) of the list.
    pub fn lpush(&mut self, db: u64, key: &[u8], values: &[&[u8]]) -> Result<(), &'static str> {
        let db_map = self.map.entry(db).or_insert_with(BTreeMap::new);

        // Check if key exists and validate type
        if let Some(value) = db_map.get(key) {
            if !matches!(value, StorageValue::List(_)) {
                return Err("WRONGTYPE Operation against a key holding the wrong kind of value");
            }
        }

        // Get or create the list
        let list = db_map
            .entry(Bytes::new(key))
            .or_insert_with(|| StorageValue::List(Vector::new()));

        // Extract the list from the StorageValue
        let list = match list {
            StorageValue::List(l) => l,
            _ => unreachable!(),
        };

        // Push values to the front in order (each one becomes the new head)
        for value in values {
            list.push_front(Bytes::new(value));
        }

        Ok(())
    }

    /// Push elements to the right (tail) of the list.
    pub fn rpush(&mut self, db: u64, key: &[u8], values: &[&[u8]]) -> Result<(), &'static str> {
        let db_map = self.map.entry(db).or_insert_with(BTreeMap::new);

        // Check if key exists and validate type
        if let Some(value) = db_map.get(key) {
            if !matches!(value, StorageValue::List(_)) {
                return Err("WRONGTYPE Operation against a key holding the wrong kind of value");
            }
        }

        // Get or create the list
        let list = db_map
            .entry(Bytes::new(key))
            .or_insert_with(|| StorageValue::List(Vector::new()));

        // Extract the list from the StorageValue
        let list = match list {
            StorageValue::List(l) => l,
            _ => unreachable!(),
        };

        // Push values to the back
        for value in values {
            list.push_back(Bytes::new(value));
        }

        Ok(())
    }

    /// Pop element from the left (head) of the list.
    pub fn lpop(&mut self, db: u64, key: &[u8]) -> Result<(), &'static str> {
        let Some(db_map) = self.map.get_mut(&db) else {
            return Ok(());
        };

        let Some(value) = db_map.get_mut(key) else {
            return Ok(());
        };

        let list = match value {
            StorageValue::List(list) => list,
            _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
        };

        if !list.is_empty() {
            *list = list.skip(1);  // Remove first element
        }

        // Remove key if list is empty
        if list.is_empty() {
            db_map.remove(key);
        }

        Ok(())
    }

    /// Pop element from the right (tail) of the list.
    pub fn rpop(&mut self, db: u64, key: &[u8]) -> Result<(), &'static str> {
        let Some(db_map) = self.map.get_mut(&db) else {
            return Ok(());
        };

        let Some(value) = db_map.get_mut(key) else {
            return Ok(());
        };

        let list = match value {
            StorageValue::List(list) => list,
            _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
        };

        if !list.is_empty() {
            list.pop_back();
        }

        // Remove key if list is empty
        if list.is_empty() {
            db_map.remove(key);
        }

        Ok(())
    }

    /// Push elements to the left (head) of the list only if list exists.
    pub fn lpushx(&mut self, db: u64, key: &[u8], values: &[&[u8]]) -> Result<(), &'static str> {
        let Some(db_map) = self.map.get_mut(&db) else {
            return Ok(());
        };

        let Some(storage_value) = db_map.get_mut(key) else {
            return Ok(());
        };

        let list = match storage_value {
            StorageValue::List(list) => list,
            _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
        };

        // Push values to the front in order (each one becomes the new head)
        for value in values {
            list.push_front(Bytes::new(value));
        }

        Ok(())
    }

    /// Push elements to the right (tail) of the list only if list exists.
    pub fn rpushx(&mut self, db: u64, key: &[u8], values: &[&[u8]]) -> Result<(), &'static str> {
        let Some(db_map) = self.map.get_mut(&db) else {
            return Ok(());
        };

        let Some(storage_value) = db_map.get_mut(key) else {
            return Ok(());
        };

        let list = match storage_value {
            StorageValue::List(list) => list,
            _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
        };

        // Push values to the back
        for value in values {
            list.push_back(Bytes::new(value));
        }

        Ok(())
    }

    /// Remove elements from list.
    /// count > 0: Remove elements equal to value moving from head to tail.
    /// count < 0: Remove elements equal to value moving from tail to head.
    /// count = 0: Remove all elements equal to value.
    pub fn lrem(&mut self, db: u64, key: &[u8], count: i64, value: &[u8]) -> Result<(), &'static str> {
        let Some(db_map) = self.map.get_mut(&db) else {
            return Ok(());
        };

        let Some(storage_value) = db_map.get_mut(key) else {
            return Ok(());
        };

        let list = match storage_value {
            StorageValue::List(list) => list,
            _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
        };

        let value_bytes = Bytes::new(value);

        if count == 0 {
            // Remove all occurrences
            *list = list.iter().filter(|v| *v != &value_bytes).cloned().collect();
        } else if count > 0 {
            // Remove from head to tail
            let max_remove = count as usize;
            let mut new_list = Vector::new();
            let mut removed = 0;
            for item in list.iter() {
                if removed < max_remove && item == &value_bytes {
                    removed += 1;
                } else {
                    new_list.push_back(item.clone());
                }
            }
            *list = new_list;
        } else {
            // Remove from tail to head (count < 0)
            let max_remove = (-count) as usize;
            let mut new_list = Vector::new();
            let mut removed = 0;
            for item in list.iter().rev() {
                if removed < max_remove && item == &value_bytes {
                    removed += 1;
                } else {
                    new_list.push_front(item.clone());
                }
            }
            *list = new_list;
        }

        // Remove key if list is empty
        if list.is_empty() {
            db_map.remove(key);
        }

        Ok(())
    }

    /// Trim list to the specified range. Both start and stop are inclusive.
    pub fn ltrim(&mut self, db: u64, key: &[u8], start: i64, stop: i64) -> Result<(), &'static str> {
        let Some(db_map) = self.map.get_mut(&db) else {
            return Ok(());
        };

        let Some(storage_value) = db_map.get_mut(key) else {
            return Ok(());
        };

        let list = match storage_value {
            StorageValue::List(list) => list,
            _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
        };

        let len = list.len() as i64;

        if len == 0 {
            return Ok(());
        }

        // Normalize negative indices
        let start_pos = if start < 0 {
            (len + start).max(0) as usize
        } else {
            start.min(len - 1).max(0) as usize
        };

        let stop_pos = if stop < 0 {
            (len + stop).max(0) as usize
        } else {
            stop.min(len - 1).max(0) as usize
        };

        // If range is invalid, delete the key
        if start_pos > stop_pos || start_pos >= len as usize {
            db_map.remove(key);
            return Ok(());
        }

        // Trim from front using skip, then trim from back to desired length
        let desired_len = stop_pos - start_pos + 1;
        *list = list.skip(start_pos).take(desired_len);

        // Remove key if list is empty
        if list.is_empty() {
            db_map.remove(key);
        }

        Ok(())
    }

    /// Insert element before or after pivot element in list.
    pub fn linsert(&mut self, db: u64, key: &[u8], before: bool, pivot: &[u8], value: &[u8]) -> Result<(), &'static str> {
        let Some(db_map) = self.map.get_mut(&db) else {
            return Ok(());
        };

        let Some(storage_value) = db_map.get_mut(key) else {
            return Ok(());
        };

        let list = match storage_value {
            StorageValue::List(list) => list,
            _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
        };

        let pivot_bytes = Bytes::new(pivot);
        let value_bytes = Bytes::new(value);

        // Find the pivot and insert
        for i in 0..list.len() {
            if list.get(i) == Some(&pivot_bytes) {
                let insert_pos = if before { i } else { i + 1 };
                list.insert(insert_pos, value_bytes);
                break;
            }
        }

        Ok(())
    }

    /// Get the length of a list.
    pub fn llen(&self, db: u64, key: &[u8]) -> Result<usize, &'static str> {
        let Some(db_map) = self.map.get(&db) else {
            return Ok(0);
        };

        let Some(value) = db_map.get(key) else {
            return Ok(0);
        };

        let StorageValue::List(list) = value else {
            return Err("WRONGTYPE Operation against a key holding the wrong kind of value");
        };

        Ok(list.len())
    }

    /// Get a range of elements from the list.
    /// Both start and stop are inclusive and support negative indices.
    pub fn lrange(&self, db: u64, key: &[u8], start: i64, stop: i64) -> Result<Vec<Bytes>, &'static str> {
        let Some(db_map) = self.map.get(&db) else {
            return Ok(Vec::new());
        };

        let Some(value) = db_map.get(key) else {
            return Ok(Vec::new());
        };

        let StorageValue::List(list) = value else {
            return Err("WRONGTYPE Operation against a key holding the wrong kind of value");
        };

        let len = list.len() as i64;

        if len == 0 {
            return Ok(Vec::new());
        }

        // Normalize negative indices
        let start_pos = if start < 0 {
            (len + start).max(0) as usize
        } else {
            start.min(len - 1).max(0) as usize
        };

        let stop_pos = if stop < 0 {
            (len + stop).max(0) as usize
        } else {
            stop.min(len - 1).max(0) as usize
        };

        if start_pos > stop_pos || start_pos >= len as usize {
            return Ok(Vec::new());
        }

        let result: Vec<Bytes> = list
            .iter()
            .skip(start_pos)
            .take(stop_pos - start_pos + 1)
            .cloned()
            .collect();

        Ok(result)
    }

    /// Set the list element at index to value.
    pub fn lset(&mut self, db: u64, key: &[u8], index: i64, value: &[u8]) -> Result<(), &'static str> {
        let Some(db_map) = self.map.get_mut(&db) else {
            return Ok(());
        };

        let Some(storage_value) = db_map.get_mut(key) else {
            return Ok(());
        };

        let list = match storage_value {
            StorageValue::List(list) => list,
            _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
        };

        let len = list.len() as i64;

        // Normalize index
        let pos = if index < 0 {
            len + index
        } else {
            index
        };

        if pos >= 0 && pos < len {
            *list = list.update(pos as usize, Bytes::new(value));
        }

        Ok(())
    }

    /// Atomically pop from the right of source and push to the left of destination.
    pub fn rpoplpush(&mut self, db: u64, source_key: &[u8], dest_key: &[u8]) -> Result<(), &'static str> {
        // Special case: same key means rotate
        if source_key == dest_key {
            let Some(db_map) = self.map.get_mut(&db) else {
                return Ok(());
            };

            let Some(value) = db_map.get_mut(source_key) else {
                return Ok(());
            };

            let list = match value {
                StorageValue::List(list) => list,
                _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
            };

            if !list.is_empty() {
                let popped = list.pop_back().unwrap();
                list.push_front(popped);
            }
            return Ok(());
        }

        // Different keys: pop from source
        let popped = {
            let Some(db_map) = self.map.get_mut(&db) else {
                return Ok(());
            };

            let Some(value) = db_map.get_mut(source_key) else {
                return Ok(());
            };

            let list = match value {
                StorageValue::List(list) => list,
                _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
            };

            if list.is_empty() {
                return Ok(());
            }

            let popped = list.pop_back().unwrap();

            // Remove source key if list is empty
            let should_delete = list.is_empty();
            if should_delete {
                db_map.remove(source_key);
            }

            popped
        };

        // Push to destination
        let db_map = self.map.entry(db).or_insert_with(BTreeMap::new);

        // Check destination type
        if let Some(value) = db_map.get(dest_key) {
            if !matches!(value, StorageValue::List(_)) {
                return Err("WRONGTYPE Operation against a key holding the wrong kind of value");
            }
        }

        let dest_list = db_map
            .entry(Bytes::new(dest_key))
            .or_insert_with(|| StorageValue::List(Vector::new()));

        let dest_list = match dest_list {
            StorageValue::List(l) => l,
            _ => unreachable!(),
        };

        dest_list.push_front(popped);

        Ok(())
    }

    /// Pop multiple elements from the left (head) of the list.
    pub fn lpop_count(&mut self, db: u64, key: &[u8], count: usize) -> Result<(), &'static str> {
        let Some(db_map) = self.map.get_mut(&db) else {
            return Ok(());
        };

        let Some(value) = db_map.get_mut(key) else {
            return Ok(());
        };

        let list = match value {
            StorageValue::List(list) => list,
            _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
        };

        // Remove up to 'count' elements from the front
        let to_remove = count.min(list.len());
        for _ in 0..to_remove {
            *list = list.skip(1);
        }

        // Remove key if list is empty
        if list.is_empty() {
            db_map.remove(key);
        }

        Ok(())
    }

    /// Pop multiple elements from the right (tail) of the list.
    pub fn rpop_count(&mut self, db: u64, key: &[u8], count: usize) -> Result<(), &'static str> {
        let Some(db_map) = self.map.get_mut(&db) else {
            return Ok(());
        };

        let Some(value) = db_map.get_mut(key) else {
            return Ok(());
        };

        let list = match value {
            StorageValue::List(list) => list,
            _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
        };

        // Remove up to 'count' elements from the back
        let to_remove = count.min(list.len());
        for _ in 0..to_remove {
            list.pop_back();
        }

        // Remove key if list is empty
        if list.is_empty() {
            db_map.remove(key);
        }

        Ok(())
    }

    /// Atomically move element from source to destination list.
    /// from_left: true = pop from left, false = pop from right
    /// to_left: true = push to left, false = push to right
    pub fn lmove(&mut self, db: u64, source_key: &[u8], dest_key: &[u8], from_left: bool, to_left: bool) -> Result<(), &'static str> {
        // Special case: same key means rotate
        if source_key == dest_key {
            let Some(db_map) = self.map.get_mut(&db) else {
                return Ok(());
            };

            let Some(value) = db_map.get_mut(source_key) else {
                return Ok(());
            };

            let list = match value {
                StorageValue::List(list) => list,
                _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
            };

            if !list.is_empty() {
                if from_left {
                    let popped = list.head().cloned().unwrap();
                    *list = list.skip(1);
                    if to_left {
                        list.push_front(popped);
                    } else {
                        list.push_back(popped);
                    }
                } else {
                    let popped = list.pop_back().unwrap();
                    if to_left {
                        list.push_front(popped);
                    } else {
                        list.push_back(popped);
                    }
                }
            }
            return Ok(());
        }

        // Different keys: pop from source
        let popped = {
            let Some(db_map) = self.map.get_mut(&db) else {
                return Ok(());
            };

            let Some(value) = db_map.get_mut(source_key) else {
                return Ok(());
            };

            let list = match value {
                StorageValue::List(list) => list,
                _ => return Err("WRONGTYPE Operation against a key holding the wrong kind of value"),
            };

            if list.is_empty() {
                return Ok(());
            }

            let popped = if from_left {
                let popped = list.head().cloned().unwrap();
                *list = list.skip(1);
                popped
            } else {
                list.pop_back().unwrap()
            };

            // Remove source key if list is empty
            let should_delete = list.is_empty();
            if should_delete {
                db_map.remove(source_key);
            }

            popped
        };

        // Push to destination
        let db_map = self.map.entry(db).or_insert_with(BTreeMap::new);

        // Check destination type
        if let Some(value) = db_map.get(dest_key) {
            if !matches!(value, StorageValue::List(_)) {
                return Err("WRONGTYPE Operation against a key holding the wrong kind of value");
            }
        }

        let dest_list = db_map
            .entry(Bytes::new(dest_key))
            .or_insert_with(|| StorageValue::List(Vector::new()));

        let dest_list = match dest_list {
            StorageValue::List(l) => l,
            _ => unreachable!(),
        };

        if to_left {
            dest_list.push_front(popped);
        } else {
            dest_list.push_back(popped);
        }

        Ok(())
    }
}
